starky/src/lookup.rs

```diff
@@ -0,0 +1,1006 @@
+/// returns the associated helper polynomials.
+pub(crate) fn get_helper_cols<F: Field>(
+    trace: &[PolynomialValues<F>],
+    degree: usize,
+    columns_filters: &[ColumnFilter<F>],
+    challenge: GrandProductChallenge<F>,
+    constraint_degree: usize,
+) -> Vec<PolynomialValues<F>> {
+    let num_helper_columns = ceil_div_usize(columns_filters.len(), constraint_degree - 1);
+
+    let mut helper_columns = Vec::with_capacity(num_helper_columns);
+
+    let mut filter_index = 0;
+    for mut cols_filts in &columns_filters.iter().chunks(constraint_degree - 1) {
+        let (first_col, first_filter) = cols_filts.next().unwrap();
+
+        let mut filter_col = Vec::with_capacity(degree);
+        let first_combined = (0..degree)
+            .map(|d| {
+                let f = if let Some(filter) = first_filter {
+                    let f = filter.eval_table(trace, d);
+                    filter_col.push(f);
+                    f
+                } else {
+                    filter_col.push(F::ONE);
+                    F::ONE
+                };
+                if f.is_one() {
+                    let evals = first_col
+                        .iter()
+                        .map(|c| c.eval_table(trace, d))
+                        .collect::<Vec<F>>();
+                    challenge.combine(evals.iter())
+                } else {
+                    assert_eq!(f, F::ZERO, "Non-binary filter?");
+                    // Dummy value. Cannot be zero since it will be batch-inverted.
+                    F::ONE
+                }
+            })
+            .collect::<Vec<F>>();
+
+        let mut acc = F::batch_multiplicative_inverse(&first_combined);
+        for d in 0..degree {
+            if filter_col[d].is_zero() {
+                acc[d] = F::ZERO;
+            }
+        }
+
+        for (col, filt) in cols_filts {
+            let mut filter_col = Vec::with_capacity(degree);
+            let mut combined = (0..degree)
+                .map(|d| {
+                    let f = if let Some(filter) = filt {
+                        let f = filter.eval_table(trace, d);
+                        filter_col.push(f);
+                        f
+                    } else {
+                        filter_col.push(F::ONE);
+                        F::ONE
+                    };
+                    if f.is_one() {
+                        let evals = col
+                            .iter()
+                            .map(|c| c.eval_table(trace, d))
+                            .collect::<Vec<F>>();
+                        challenge.combine(evals.iter())
+                    } else {
+                        assert_eq!(f, F::ZERO, "Non-binary filter?");
+                        // Dummy value. Cannot be zero since it will be batch-inverted.
+                        F::ONE
+                    }
+                })
+                .collect::<Vec<F>>();
+
+            combined = F::batch_multiplicative_inverse(&combined);
```

Comment:
Just a thought: this inverting function seems to be called often. I wonder if it could be a target for SIMD optimization? It'd be a bit of a challenge, but perhaps worth it at some point?